### PR TITLE
Add saml interface

### PR DESCRIPTION
**Pull request by arturo-seijas**
Add SAML interface and saml-integrator charm as a provider

### CODE REVIEW

**Review by gtrkiller — APPROVED** [review]
LGTM

File: docs/json_schemas/saml/v0/provider.json [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        "metadata_url",
+        "entity_id",
+        "single_sign_on_service_redirect_url",
+        "single_sign_on_service_redirect_binding",
+        "single_sign_on_service_post_binding",
+        "x509certs"
+      ]
+    }
+  }
+}
```
Comment (yanksyoon) [inline docs/json_schemas/saml/v0/provider.json:154]: I think a newline is missing here :D

File: docs/json_schemas/saml/v0/requirer.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    }
+  },
+  "definitions": {
+    "BaseModel": {
+      "title": "BaseModel",
+      "type": "object",
+      "properties": {}
+    }
+  }
+}
```
Comment (yanksyoon) [inline docs/json_schemas/saml/v0/requirer.json:20]: And here

File: interfaces/saml/v0/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# `saml`
+
+## Overview
+
+This relation interface describes the expected behavior between of any charm claiming to be able to interface with a SAML Provider and the SAML Provider itself.
+
+## Usage
+
+In most cases, this will be accomplished using the [saml library](), although charm developers are free to provide alternative libraries as long as they fulfil the behavioural and schematic requirements described in this document.
```
Comment (yanksyoon) [inline interfaces/saml/v0/README.md:9]: I think you might've wanted to put a link here? [saml library]()

**Review by yanksyoon — APPROVED** [review]
LGTM, minor comments!

**Comment by simskij** [comment]
@massigori, i would love your input here before we merge this.

**Comment by massigori** [comment]
@simskij we already discussed this interface in ISD058. This interface is supposed to support internally adopted charms that require Launchpad group information.

@arturo-seijas I think you should amend the README to specify the intended use case as the properties are not sufficient to integrate with some SAML identity providers that require custom attributes.

**Comment by arturo-seijas** [comment]
> @simskij we already discussed this interface in ISD058. This interface is supposed to support internally adopted charms that require Launchpad group information.
> 
> @arturo-seijas I think you should amend the README to specify the intended use case as the properties are not sufficient to integrate with some SAML identity providers that require custom attributes.

Done @massigori @simskij

File: interfaces/saml/v0/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+This relation interface describes the expected behavior between of any charm claiming to be able to interface with a SAML Provider and the SAML Provider itself. This interface has been defined to support charms that require Launchpad group information. Other Identity Providers can be used interchangeably as long as custom attributes are not required.
+
+## Usage
+
+In most cases, this will be accomplished using the [saml library](https://github.com/canonical/saml-integrator-operator/blob/main/lib/charms/saml_integrator/v0/saml.py), although charm developers are free to provide alternative libraries as long as they fulfil the behavioural and schematic requirements described in this document.
+
+## Direction
+
+The `saml` interface implements a provider pattern.
```
Comment (gruyaume) [inline interfaces/saml/v0/README.md:13]: What is a provider pattern? Don't we mean `provider/requirer`?

File: interfaces/saml/v0/schema.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
+# Copyright 2023 Canonical
+# See LICENSE file for licensing details.
+"""This file defines the schema for the provider side of the saml interface.
+
+It exposes ones interfaces.schema_base.DataBagSchema subclass called:
```
Comment (gruyaume) [inline interfaces/saml/v0/schema.py:5]:
```suggestion
It exposes one interfaces.schema_base.DataBagSchema subclass called:
```

File: interfaces/saml/v0/README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+#### Example
+
+```yaml
+related-units: {}
+application_data: {
+  "metadata_url": "https://login.ubuntu.com/saml/metadata",
+  "entity_id": "https://login.ubuntu.com",
+  "single_sign_on_service_redirect_url": "https://login.ubuntu.com/saml/",
+  "single_sign_on_service_redirect_binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect",
+  "single_logout_service_redirect_url": "https://login.ubuntu.com/+logout",
```
Comment (gruyaume) [inline interfaces/saml/v0/README.md:54]: For the redirect url and binding the prefix is `single_sign_on` and here it's `single`, we should use the same for both

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: interfaces/saml/v0/schema.py (hunk 5 above)
Comment (arturo-seijas) [reply interfaces/saml/v0/schema.py]: Fixed. Thx

File: interfaces/saml/v0/README.md (hunk 4 above)
Comment (arturo-seijas) [reply interfaces/saml/v0/README.md]: Fixed. Thx

File: interfaces/saml/v0/README.md (hunk 6 above)
Comment (arturo-seijas) [reply interfaces/saml/v0/README.md:54]: The prefixes are "single_singn_on" and "single_logout". These have already been agreed as part of a spec

**Review by gruyaume — APPROVED** [review]

**Pull request merged**